### PR TITLE
bpo-30058: Fixed buffer overflow in select.kqueue.control().

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue30058

### CODE REVIEW

File: Modules/selectmodule.c [hunk 1]
```diff
@@ -2148,37 +2148,31 @@ kqueue_queue_control(kqueue_queue_Object *self, PyObject *args)
...
-                            "changelist is not iterable");
+        seq = PySequence_Fast(ch, "changelist is not iterable");
+        if (seq == NULL) {
             return NULL;
         }
-        nchanges = PyObject_Size(ch);
-        if (nchanges < 0) {
-            goto error;
-        }
+        nchanges = PySequence_Fast_GET_SIZE(arg);
```
Comment (zhangyangyu) [inline Modules/selectmodule.c]: `PySequence_Fast_GET_SIZE` returns a Py_ssize_t. This may trigger a compile warning.

Comment (serhiy-storchaka) [reply Modules/selectmodule.c]: Good point!

File: Modules/selectmodule.c [hunk 2]
```diff
@@ -2148,37 +2148,36 @@ kqueue_queue_control(kqueue_queue_Object *self, PyObject *args)
...
-        if (it == NULL) {
-            PyErr_SetString(PyExc_TypeError,
-                            "changelist is not iterable");
+        seq = PySequence_Fast(ch, "changelist is not iterable");
+        if (seq == NULL) {
             return NULL;
         }
-        nchanges = PyObject_Size(ch);
-        if (nchanges < 0) {
+        if (PySequence_Fast_GET_SIZE(arg) > INT_MAX / sizeof(struct kevent)) {
```
Comment (zhangyangyu) [inline Modules/selectmodule.c]: `arg` should be `seq`, right?

I don't think we need to `/ sizeof(struct kevent)`. `> INT_MAX` is enough. Then there will be a memory error below. Or use `PY_SSIZE_T_MAX` here, assuming `PY_SSIZE_T_MAX / sizeof(struct kevent) <= INT_MAX`.

Comment (serhiy-storchaka) [reply Modules/selectmodule.c]: Done.

**Review by zhangyangyu — APPROVED** [review]
LGTM from code. But I can't even compile.

File: Lib/test/test_kqueue.py [hunk 3]
```diff
@@ -208,6 +208,31 @@ def testPair(self):
         b.close()
         kq.close()
 
+    def test_issue30058(self):
+        # chagelist must be an iterable
```
Comment (louisom) [inline Lib/test/test_kqueue.py]: typo: `chagelist` -> `changelist`

**Review by louisom — CHANGES_REQUESTED** [review]
just a nit typo

**Comment by louisom** [comment]
@zhangyangyu why can't you compile? My env compile it well.

**Comment by zhangyangyu** [comment]
@lulouie , the code is specialized for BSD. I only have access to Linux. Compiling it on Linux doesn't make any sense.

**Comment by louisom** [comment]
@zhangyangyu ah, It has compiled on Linux, but only the GNU part, not BSD part.

File: Lib/test/test_kqueue.py (hunk 3 above)
Comment (serhiy-storchaka) [reply Lib/test/test_kqueue.py]: Good catch @lulouie!

**Pull request merged**